core/app/models/spree/promotion_handler/coupon.rb

```diff
@@ -17,6 +17,11 @@ def apply
             handle_present_promotion(promotion)
           elsif promotion_code && promotion_code.promotion.inactive?
             set_error_code :coupon_code_expired
+          # If the promotion code exists but has no actions,
+          # then log warning and show not found error to user
```

Comment:
I think it's better avoid adding code comments unless strictly needed. What happens is already clear by reading the code.